Fix standard_hq_report

orangejenny opened the pull request:
## Product Description
https://dimagi.atlassian.net/browse/SAAS-16512

Fixes a bug where UCRs could not be saved as saved reports.

## Technical Summary
Introduced in https://github.com/dimagi/commcare-hq/pull/35599, specifically the removal of the block [here](https://github.com/dimagi/commcare-hq/pull/35599/files#diff-c140afd2d0403854fb9c2d60bda41f8409d03f856c2c46cd080e7aafefcc397dL36-L38) that handled UCRs by calling out to the UCR version of `getStandardReport`.

`getStandardReport` was split into a UCR version and a standard reports version many years ago, because UCRs were migrated to requirejs and standard reports were not. Now that all reports are on webpack, this PR merges the two sets of logic back into one. The UCR version included some UCR-specific analytics code that's staying in the UCR module.

This means updating UCRs to use one big `js_options` object for report options, the way standard reports do, instead of passing the report options as individual pieces of data. Because `js_options` is built in python, this highlighted that several of these options have never done anything in UCRs. Additional context on this...

Standard reports all derive from `GenericReportView`, which sets up a `report` object in the context [here](https://github.com/dimagi/commcare-hq/blob/bba4cfffc632c956628539be42669e18fe41cebe/corehq/apps/reports/generic.py#L536).

In UCRs, the `report` in the context is the view itself (see [here](https://github.com/dimagi/commcare-hq/blob/bba4cfffc632c956628539be42669e18fe41cebe/corehq/apps/userreports/reports/view.py#L347)), which has some inconsistencies with `GenericReportView`.

Expressions like `report.filter_set`, which don't exist in the UCR view class, were previously failing silently when they got called during template rendering and are now removed. This also made it apparent that although UCRs have `emailable` set to True, the initial page data incorrectly referenced `report.is_emailable` and therefore the functionality wasn't working. I started to implement emailability for UCRs, since clearly that was the original intention, but I backed out after realizing that it would be more involved than expected.

This PR does, however, turn on the export all functionality, which was similarly previously broken but now should work because `isExportAll` is now based on the [real](https://github.com/dimagi/commcare-hq/blob/bba4cfffc632c956628539be42669e18fe41cebe/corehq/apps/userreports/reports/view.py#L152) `exportable_all` view attribute, instead of `is_export_all` which exists in the context of standard reports (see [here](https://github.com/dimagi/commcare-hq/blob/bba4cfffc632c956628539be42669e18fe41cebe/corehq/apps/reports/generic.py#L556)) but doesn't exist in UCRs.

This PR has made me much less fond of renames [like this](https://github.com/dimagi/commcare-hq/blob/bba4cfffc632c956628539be42669e18fe41cebe/corehq/apps/reports/generic.py#L555-L556) where the name changes but the meaning does not. Not that I was fond of them in the first place.

## Feature Flag
UCRs

## Safety Assurance

### Safety story
I'm testing on staging. I'm leaning towards requesting QA to make sure I didn't break any other basic reports interaction in either standard reports or UCRs.

I've tested the inddex reports on staging.

### Automated test coverage

No

### QA Plan

https://dimagi.atlassian.net/browse/QA-7456

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Combined standard report logic and UCR logic

This combines the UCR version of getStandardHQReport, which the breaking change caused to no longer get called,
into the logic in standard_hq_report.js. It sets up a js_options dictionary to use in UCR, which was previously
using initial page data.

It also eliminates the old block for handling custom reports, which should be irrelevant now that all custom
reports use external js files and hqDefine, instead of ineline js and global variables.

orangejenny committed:
Removed url_root

This isn't necessary for UCRs. It's only used in getReportBaseUrl, which is used in getReportRenderUrl
(which UCR overrides) and sendEmail (which doesn't currently matter because UCR emailing is broken for other
reasons).

orangejenny committed:
Removed attributes that were being ignored amyway

orangejenny committed:
Updated B5

jingcheng16 commented:
@orangejenny Hi Jenny, do you mind showing me the code that prove this? I trust you, I'm just curious. I searched `StandardReportView` (my guess of the view) and didn't find anything 😂 
> Standard reports all derive from GenericReportView. 

Also, I saw you're adding another `js_options` to `ConfigurableReportView`. This make sense to me. So both standard report and ucr will have different `js_options`. What confused me is, when I tried to fix it, I simply add `initialPageData.get('js_options')` in ucr's report options, without adding `js_options` in `ConfigurableReportView`. It works. But after seeing your fix, I'm surprised that m fix works locally, though the `yarn build` failed...

jingcheng16 commented:
>'require' is not defined

Want to ask about this lint error too. It is unrelated to your change. Just curious what is it. I noticed we have migrated `hqRequire` to be something else (I forgot). I wonder if this `require` is a similar thing that need migration.

orangejenny committed:
Switched ConfigurableReportView.exportable_all to False

This is to restore previous behavior. The old code had exportable_all=True but ignored the value.

jingcheng16 commented:
Thanks Jenny.
I read your reply and plan to respond but got distracted last week.
Here is my branch: jc/fix-configurable-report. I have standard_hq_report to always call configurable_report, so I guess configurable report will get the `js_option` from there.

jingcheng16 reviewed (approved):
LGTM! Thank you for your explanation.

orangejenny commented:
@coderabbitai review

orangejenny commented:
@coderabbit review

orangejenny commented:
@jingcheng16 Ah, I think I get what's going on in your branch versus mine.

The `initialPageData.get('js_options')` you added doesn't do anything (I checked out your branch locally to confirm, it's undefined). But since you're still calling the UCR version of `getStandardHQReport`, that code is still looking up all the individual pieces of initial page data ([these lines](https://github.com/dimagi/commcare-hq/compare/master...jc/fix-configurable-report?w=1#diff-90636777d2ce3183e452160ad6c6e09ff0bc9567618752ba1950ee56ddafb9c7R52-R54) and the ones after them).

Since I deleted `getStandardHQReport` from `configurable_report.js` and switched to calling `getStandardReport`/`getStandard` in `standard_hq_report.js`, that logic is expecting `js_options`.

orangejenny committed:
Removed return value from configurable_report.js

This is something RequireJS used to want, for modules to return a truthy value.

jingcheng16 commented:
@orangejenny Thank you for your explanation. 
So in my version, the `js_options` is `underfined`, but it will correctly get `type` from `initialPageData`, thus the code for ucr works?

Can approve again once the test is fixed. Test failure seems due to your last commit.

orangejenny commented:
@jingcheng16 Yes, I believe so.

I think the test failure is just a flaky one, let's see if re-running helps.

jingcheng16 reviewed (approved):
Nice. All tests passed.
